packages/react-router/modules/hooks.js

```diff
@@ -44,12 +44,17 @@ export function useParams() {
 export function useRouteMatch(path) {
   if (__DEV__) {
     invariant(
-      typeof useContext === "function",
+      typeof useContext === "function" && typeof useMemo === "function",
       "You must use React >= 16.8 in order to use useRouteMatch()"
     );
   }
 
   const location = useLocation();
   const match = useContext(RouterContext).match;
-  return path ? matchPath(location.pathname, path) : match;
+  const pathMatch = useMemo(
+    () => (path ? matchPath(location.pathname, path) : null),
+    [location.pathname, path]
+  );
```

Comment:
I'm not 100% sure what your use case is here so ignore me if it's not relevant, but I'm not sure `useMemo` is safe if you're wanting to perform change detection? -- https://github.com/ReactTraining/react-router/issues/7059#issuecomment-628783747

`useMemoOne` offers a stable cache however -- https://github.com/alexreardon/use-memo-one

Reply:
Yes, you are correct that `useMemo` is not a semantic guarantee that this value will be preserved between renders. As the React documentation states:

> **You may rely on useMemo as a performance optimization, not as a semantic guarantee.** In the future, React may choose to “forget” some previously memoized values and recalculate them on next render, e.g. to free memory for offscreen components. Write your code so that it still works without `useMemo` — and then add it to optimize performance.

However given the current implementation of `useMemo` is basically a 'remember the last version' cache, which is exactly what we need for this specific case. I've seen `useMemo` being used as a 'stable cache' (given the last value) in the wild, and the fact that the React team discourages this without providing a easy to use stable alternative is surprising to me.

If the React team would ever want to change this they would have to make it a breaking change in a major version, so right now this doesn't seem like a problem to me. That said, this code could possibly be rewritten to make use of a Ref instead. It's a shame we'd have to include another dependency just to get a stable memo.

Reply:
@jonkoops ha yes that's [largely my thinking too](https://github.com/facebook/react/issues/15278), I don't know why React isn't either removing that from the docs or providing a solution.

Reply:
FYI it looks like there are already several places in the `dev` branch where `useMemo()` is used in places where stability might be expected.